### DISCUSSION TITLE
Test file location

### DISCUSSION THREAD

**Discussion by fschade** [owncloud/web#5425] [General]
In Web currently we use two places where tests are located.
- right beside the src folder 
- right beside the file to test

We need to discuss and then decide what location fits best.

I'd like to propose following:
```
.
├── __mocks__ (mocks for node_modules)
│   └── axios.js (mock for node_modules/axios)
├── node_modules
│   └── axios
└── packages
    └── web-app-files
        └── src
            └── helpers
                └── resource
                    ├── __mocks__ (mocks for resource module)
                    │   └── path.js (mock path.js)
                    ├── __tests__ (tests for entire resource module)
                    │   └── batchActions.spec.js (concrete unit test for batchActions.js from resource module)
                    ├── batchActions.js
                    ├── index.js
                    └── path.js
```

## this brings following benefits:
- [jest manual mocks](https://jestjs.io/docs/manual-mocks)
- easier mental model where tests belong to
- we test from child folder to parent folder and can test also non exported files (by index.js)

## other projects that do it that way:
- https://github.com/facebook/jest/tree/master/packages/jest-core/src/lib
- https://github.com/facebook/react/tree/main/packages/react-client/src

# history:
- [discussion](https://talk.owncloud.com/channel/web?msg=MFv2qHbfatwaDfCtW)
- https://github.com/owncloud/web/pull/5412
- https://github.com/owncloud/web/pull/5194
- https://github.com/owncloud/web/pull/5224

# misc
- where to put integration tests?

**Comment by fschade** [comment]
cC.: @LukasHirt, @pascalwengerter, @kulmann, @individual-it, @phil-davis

**Comment by individual-it** [comment]
Just to point it out: this is only about unit tests, there are also integration and acceptance tests in web, that would still stay in the `tests` folder

And that is also the only slight disadvantage I can think of the proposal. Will people be confused to find some tests (integration/acceptance)  in the `tests` folder and unit tests not? But maybe that is not an issue at all.

generally I don't have any strong opinions about where the unit tests should live, as long as its uniform across the project

we would need to add these items into the model above:
- config files, currently living in `/tests/unit/config`, they cannot go into the root folder because the integration tests are also run with jest and have their own config files
- stubs, currently under `/tests/unit/stubs`. Can they go into the main `__mocks__` folder?
- helpers, currently under `/tests/unit/helpers`

**Comment by fschade** [comment]
another point that came to my mind a second ago while writing tests.
If the tests located in the folder beside the src, it's problematic to mock other modules because the path can be pretty obscure.

```
jest.mock('../../../../src/helpers/resource/publicPreviewUrl', mock)
```

**Reply by LukasHirt** [reply]
That's what we can use mappers for. `@/` for root, `@files/` for files app root, etc. No matter the decision on the location of tests I think those makes sense

**Comment by LukasHirt** [comment]
I have the same thoughts as @individual-it regarding having 2/3 tests in dedicated folder and rest next to files. What adds a bit to the confusion then is again as pointed out already that the `unit` folders still exists in the test folders but wouldn't then contain specs resulting in feeling like "there aren't any". Testing non-exported functions is of course a huge benefit though.

What I am also considering is, how do we then make clear separation what is integration test and what is unit test? Integration test behave a lot like unit tests - they are directly tied to a component. They are also called `*.spec.js`. They are both running with jest and share 90 % of the config (that would be actually nice to have a shared file which is then merged into specific config...). They can share fixtures, mocks, etc. So I am a bit afraid of the confusion of which one is which.

**Comment by individual-it** [comment]
what to do next with this discussion? 

CC @pascalwengerter @kulmann

**Comment by pascalwengerter** [comment]
We've decided to go with a dedicated `tests/` folder containing respective sub-folders for `unit`, `integration` etc on a package(s) level and will adjust the repo accordingly

**Reply by fschade** [reply]
https://github.com/owncloud/web/issues/5471

**Reply by individual-it** [reply]
should be fixed by #5412
